### PR TITLE
Return subproject source root when asking for meson.source_root() in …

### LINKED ISSUE #722: [regression] Error computing .def files location when using subprojects
Since:

```
commit 88037c43858c600d5004ede1ab01378cb110c8af
Author: Nirbheek Chauhan <nirbheek@centricular.com>
Date:   Thu Aug 18 20:52:22 2016 +0530

    shared_library: Add vs_module_defs to link_depends

    With this, if the module definitions file is edited, the shared library
    and all reverse-dependencies will be re-linked.

    Fixes #643

:040000 040000 84e77a840d469fc4e4821733e18542e92f0dfaa5 c7ea8aca66e8e4150e23966e75ae525375f8f330 M  mesonbuild
```

I am getting the following error:

```
ninja: Entering directory `mesonbuild'
ninja: error: '/home/thiblahute/devel/gstreamer/master/gst-all/win32/common/libgstreamer.def', needed by 'subprojects/gstreamer/gst/libgstreamer-1.0.so.0.901.0', missing and no known rule to make it
ninja: Entering directory `mesonbuild'
ninja: error: '/home/thiblahute/devel/gstreamer/master/gst-all/win32/common/libgstreamer.def', needed by 'subprojects/gstreamer/gst/libgstreamer-1.0.so.0.901.0', missing and no known rule to make it
```

Trying to build [gst-all](https://github.com/thiblahute/gst-all)

### PR DESCRIPTION
**Pull request by jpakkane**
…a subproject. Closes #722.

### CODE REVIEW

File: mesonbuild/interpreter.py [hunk 1]
```diff
@@ -943,7 +943,11 @@ def current_build_dir_method(self, args, kwargs):
...
 
     def source_root_method(self, args, kwargs):
-        return self.interpreter.environment.source_dir
+        if self.interpreter.subproject == '':
+            return self.interpreter.environment.source_dir
+        return os.path.join(self.interpreter.environment.source_dir,
+                            self.interpreter.subproject_dir,
+                            self.interpreter.subproject)
 
     def build_root_method(self, args, kwargs):
```
Comment (nirbheek) [inline mesonbuild/interpreter.py]: Absolutely. We should also set `MESON_SOURCE_ROOT` and `MESON_BUILD_ROOT` per-subproject somehow. Right now those are also wrong for subprojects afaict.

**Comment by nirbheek** [comment]
I think the issue is that `do_subproject()` reuses `self.build`. We should reconstruct it so that it also has a new environment with the root at the subproject level. That way we won't need to add all those hacks and everything should Just Work(TM).

**Comment by tp-m** [comment]
I don't think you can make the same assumptions about the build directory, what about flat builds and such?

**Comment by jpakkane** [comment]
Yeah, the build dir can be totally different from source tree. If it were up to me we would not even have a function for top level build dir, but it's not really practical.

**Comment by nirbheek** [comment]
You misunderstand. `self.build` has the build information, including the environment in `self.build.environment` which contains information about the source tree and the build root, etc. That's being copied over instead of being instantiated anew.

**Comment by jpakkane** [comment]
This was for the mention above:

> Should we have the same behaviour for build root?

So what should `meson.build_root()` return for subprojects? Is it the topmost build root dir or `subprojects/foobar`? The latter has the problem that different backends have different layouts for the builddir and Ninja build, too, since you can set `layout=flat`.

**Comment by jpakkane** [comment]
The `self.build` thing was something I pondered for a long time. The problem is that if you have a separate build object, then you need to write a build object merger. Currently it does the merges immediately (because they are in the same `build` variable). There are pros and cons to each approach.

**Comment by jpakkane** [comment]
Is this ok to merge now? We can fix the build dir later if needed.

**Comment by nirbheek** [comment]
I still think that the cleaner solution is to have subprojects be invoked in a more “recursive” way and to shovel whatever ugliness is required into the merging rather than pepper exceptions across the codebase.

This commit only fixes one instance of the source root being incorrect, does it not? Afaict, the `MESON_SOURCE_ROOT` env variable is also incorrect in subproject command executions.

**Comment by jpakkane** [comment]
Gaaaah! I wish I had never put source root in. Current source dir is what you need almost always.

**Comment by thiblahute** [comment]
> Gaaaah! I wish I had never put source root in. Current source dir is what you need almost always.

As we need it, it probably means it was a good idea to put it ;)

**Comment by jpakkane** [comment]
I'm going to be inside an airplane for hours this weekend. I'll see if I can do the refactoring necessary to make this happen.

**Comment by jpakkane** [comment]
Well as expected I had to focus on a bunch of other stuff instead. Hopefully I'll get a chance at the beginning of this week. However thinking about it this might not be as straightforward as thought. It is unclear if putting the merge in one place cleans things up sufficiently as it will have its own ugliness. The only way is to try, I guess.

**Comment by jpakkane** [comment]
@nirbheek should we merge this? It is not (arguably) the correct solution but at least it gives the behaviour we expect. Maybe add a test first.

**Comment by nirbheek** [comment]
I think we should do the same for build_root (taking care of flat layouts), and also fix the environment variables. The build manager merge can happen later if it's needed.

And yeah, need tests...

**Comment by thiblahute** [comment]
I am afraid this is gonna break builds now tbh.

**Comment by nirbheek** [comment]
This only breaks people building with subprojects, but it would still break building stable gstreamer, so that sucks. I had an idea for how to do transitions though, so maybe we can try it out with this.

Idea was that we check the `meson_version` specified in `project()`, and if it is older than the version in which the change was made, we use the older behaviour and print a warning. This code would be kept around for at most 3 releases (so people have 3 months) and then removed.

**Comment by jpakkane** [comment]
Seems reasonable. We are probably going to have a couple of deprecations so it might make sense to come up with some support code for it.

**Comment by jpakkane** [comment]
> Idea was that we check the meson_version specified in project()

But the problem here is that most projects don't specify it so we don't know what version they were written for.

**Comment by nirbheek** [comment]
> But the problem here is that most projects don't specify it so we don't know what version they were written for.

My implementation for this in a PR that I am too lazy to find now was that specifying no version implies that the user wants the deprecated behaviour + warning. I think this is reasonable, and also what people expect.

**Comment by jpakkane** [comment]
Maybe we could add a kwarg to specify which dir the user wants?

**Comment by nirbheek** [comment]
I would agree re: kwarg if subprojects were seeing a lot of use, but so far our biggest users are gst-build and copylibs within the GNOME project, and only gstreamer depends on this behaviour afaik, so this is a good time to set a good default.

Once we start gathering steam on Windows (I'm working on enabling that for gstreamer and gnome) it'll be much harder.

**Comment by nirbheek** [comment]
Okay. Let's drop this. I'm already seeing many projects which assume `meson.source_root()` is the source root, and not the subproject source root.

I propose that we add new methods called `meson.project_source_root()` + `meson.project_build_root()`.

**Pull request closed without merging**